compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -594,6 +597,55 @@ trait Checking {
       ctx.error(ex"$cls cannot be instantiated since it${rstatus.msg}", pos)
   }
 
+  /** Check that pattern `pat` is irrefutable for scrutinee tye `pt`.
+   *  This means `pat` is either marked @unchecked or `pt` conforms to the
+   *  pattern's type. If pattern is an UnApply, do the check recursively.
+   */
+  def checkIrrefutable(pat: Tree, pt: Type, isPatDef: Boolean)(implicit ctx: Context): Boolean = {
+
+    def check(pat: Tree, pt: Type): Boolean = {
+      if (pt <:< pat.tpe)
+        true
+      else {
+        var reportedPt = pt.dropAnnot(defn.UncheckedAnnot)
+        if (!pat.tpe.isSingleton) reportedPt = reportedPt.widen
+        val problem = if (pat.tpe <:< reportedPt) "is more specialized than" else "does not match"
+        val fix = if (isPatDef) "`: @unchecked` after" else "`case ` before"
+        ctx.errorOrMigrationWarning(
```

Comment:
I think we should either make this a warning, or turn the corresponding pattern matching warning into type error.

Reply:
I'm actually all for making inexhaustive match an error and requiring an `@inexhaustive` annotation to accept it.